src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -356,73 +349,60 @@ internal static void RunFromThreadPoolDispatchLoop(ExecutionContext executionCon
-                if (executionContext.HasChangeNotifications)
-                {
-                    // There are change notifications; trigger any affected
-                    OnValuesChanged(previousExecutionCtx: null, executionContext);
-                }
+                OnValuesChanged(previousExecutionCtx: null, executionContext);
             }
 
-            // Note: we aren't running in try/catch as if an exception is directly thrown on the threadpool
-            // either process will crash or its a ThreadAbortException. 
-            // In either case sending AsyncLocal change notifications is unimportant.
             callback.Invoke(state);
 
             ExecutionContext currentExecutionCtx = currentThread.ExecutionContext;
-            if (currentExecutionCtx != null)
+            if (currentExecutionCtx.HasChangeNotifications)
             {
-                // Reset ExecutionContext to null (Default) prior to calling OnValuesChanged callback
+                // Restore to Default before Notifications
                 currentThread.ExecutionContext = null;
-                if (currentExecutionCtx.HasChangeNotifications)
-                {
-                    // There are change notifications; trigger any affected
-                    OnValuesChanged(currentExecutionCtx, nextExecutionCtx: null);
-                }
+                OnValuesChanged(currentExecutionCtx, nextExecutionCtx: null);
             }
 
-            // For QUWICallbacks there are no extra steps and ThreadPoolWorkQueue.Dispatch 
-            // will reset EC and SyncCtx back to defaults, so it doesn't need to be done here.
+            // ThreadPoolWorkQueue.Dispatch will reset EC and SyncCtx back to default
         }
 
         internal static void RunForThreadPoolUserWorkItemWithDefaultContext<TState>(Action<TState> callback, in TState state)
         {
-            Debug.Assert(Thread.CurrentThread.IsThreadPoolThread);
-            Debug.Assert(Thread.CurrentThread.ExecutionContext == null, "ThreadPool thread not on Default ExecutionContext.");
-            Debug.Assert(Thread.CurrentThread.SynchronizationContext == null, "ThreadPool thread not on Default SynchronizationContext.");
+            CheckThreadPoolAndContextsAreDefault();
 
-            // Note: we aren't running in try/catch as if an exception is directly thrown on the threadpool
-            // either process will crash or its a ThreadAbortException. 
-            // In either case sending AsyncLocal change notifications is unimportant.
             callback.Invoke(state);
 
-            // Capture current thread to local rather than looking it up multiple times
             Thread currentThread = Thread.CurrentThread;
             ExecutionContext currentExecutionCtx = currentThread.ExecutionContext;
-            if (currentExecutionCtx != null)
+            if (currentExecutionCtx != null && currentExecutionCtx.HasChangeNotifications)
```

Comment:
If the current non-default context does not have notifications it would not be reset to null, but unless Dispatch nulls it after this work item completes, would it not be possible for a timer callback to observe a non-null context on start?

Reply:
Which I guess is ok because unmanaged work items don't reset either, nevermind. In that case though, why reset the thread's execution context to null at all in `RunFromThreadPoolDispatchLoop` and `RunForThreadPoolUserWorkItem`?

Reply:
For both they set back to the original prior to the change notifications running; and for the `RunFromThreadPoolDispatchLoop` since the more async machinery runs after; it has to be set back before it returns.

There is even an AsyncLocal test that mutates the EC in the change handler (but changing an asynclocal) which it then checks leaks out of the await.

Reply:
The differences seem to be subtle, I'm not sure it's worth distinguishing those. Why not keep it simple and always reset the execution context? It'll be an extra write but it'll keep the methods symmetric and it wouldn't require extra care to use them. If we really want to optimize out the write I would suggest the opposite, and not reset the execution context in any of these methods, and leave that to the caller for all cases (with a comment).

Reply:
Resetting before notifications I guess is necessary as it is the previous behavior, but when notifications are not necessary

Reply:
Yeah, so it doesn't reset unless there are notifications; because the current context can be observed in the notification handler. Except in the non UserWorkItem case where it always resets because they don't return directly to the Dispatch loop (e.g. Task) so again the context will be observed (and passed out of any awaits)

Reply:
Done